File: sdk-python/synapse/llm/bridges.py
"""Bridges from popular vendor clients to Synapse's InferenceAdapter.

Each ``from_*`` accepts the vendor's existing client object (so the user
keeps their model choice, API key, base URL, retry/timeout config, etc.)
and returns an adapter Synapse can use.

These bridges are LIGHTWEIGHT: they reuse the existing Synapse adapters
(``AnthropicAdapter``, ``OpenAIAdapter``, ``GeminiAdapter``, ``OllamaAdapter``)
where possible, just constructed from the user's already-configured
client instead of from environment variables.

Lazy imports throughout — Synapse never depends on ``langchain`` or
``litellm`` unless the user explicitly bridges through them.
"""
from __future__ import annotations

import logging
import os
from typing import Any, Optional

from synapse.adapters.base import InferenceAdapter

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Anthropic
# ---------------------------------------------------------------------------
def from_anthropic(
    client: Optional[Any] = None,
    *,
    model: str = "claude-haiku-4-5-20251001",
    api_key: Optional[str] = None,
) -> InferenceAdapter:
    """Wrap an Anthropic client (or create one from env) for Synapse.

    Examples:
        from anthropic import AsyncAnthropic
        synapse.set_llm(synapse.from_anthropic(AsyncAnthropic()))

        # Or skip the client and let Synapse build one
        synapse.set_llm(synapse.from_anthropic(model="claude-haiku-4-5-20251001"))
    """
    from synapse.adapters.hosted.anthropic_adapter import AnthropicAdapter

    adapter = AnthropicAdapter(model=model, api_key=api_key)
    if client is not None:
        # If the user passed a configured client, swap it in. This preserves
        # their retry config, base_url, default_headers, etc.
        adapter._client = client
    return adapter


# ---------------------------------------------------------------------------
# OpenAI
# ---------------------------------------------------------------------------
def from_openai(
    client: Optional[Any] = None,
    *,
    model: str = "gpt-4o-mini",
    api_key: Optional[str] = None,
) -> InferenceAdapter:
    """Wrap an OpenAI / Azure OpenAI / OpenAI-compatible client.

    Examples:
        from openai import AsyncOpenAI
        synapse.set_llm(synapse.from_openai(AsyncOpenAI()))

        # Self-hosted vLLM via OpenAI-compatible endpoint
        client = AsyncOpenAI(base_url="http://localhost:8000/v1", api_key="sk-local")
        synapse.set_llm(synapse.from_openai(client, model="llama3.1-8b"))
    """
    from synapse.adapters.hosted.openai_adapter import OpenAIAdapter

    adapter = OpenAIAdapter(model=model, api_key=api_key)
    if client is not None:
        adapter._client = client
    return adapter


# ---------------------------------------------------------------------------
# LangChain / LangGraph
# ---------------------------------------------------------------------------
def from_langchain(llm: Any) -> InferenceAdapter:
    """Wrap a LangChain ``BaseLanguageModel`` so Synapse can call it.

    Works with any LangChain-compatible chat model: ``ChatAnthropic``,
    ``ChatOpenAI``, ``ChatGroq``, ``ChatVertexAI``, ``ChatOllama``, etc.
    Reuses the user's existing model+config.

    Example:
        from langchain_anthropic import ChatAnthropic
        synapse.set_llm(synapse.from_langchain(ChatAnthropic(model="...")))
    """
    return _LangChainBridgeAdapter(llm)


# ---------------------------------------------------------------------------
# LiteLLM (universal across 100+ providers)
# ---------------------------------------------------------------------------
def from_litellm(
    *,
    model: str,
    api_key: Optional[str] = None,
    api_base: Optional[str] = None,
    **extra: Any,
) -> InferenceAdapter:
    """Wrap a LiteLLM-routable model spec (e.g. "anthropic/claude-haiku-4-5").

    LiteLLM normalizes 100+ providers behind one client. Synapse stays
    single-dependency by lazy-importing it only when this bridge is used.

    Example:
        synapse.set_llm(synapse.from_litellm(model="anthropic/claude-haiku-4-5"))
    """
    return _LiteLLMBridgeAdapter(model=model, api_key=api_key, api_base=api_base, extra=extra)


# ---------------------------------------------------------------------------
# Auto-detect (best-effort) — picks the first provider it finds keys for
# ---------------------------------------------------------------------------
def auto_llm() -> InferenceAdapter:
    """Pick the cheapest available provider based on env keys.

    Order: Anthropic → OpenAI → Gemini → Ollama (local).
    Picks the cheapest model in that family by default.

    Raises:
        RuntimeError if no provider can be configured.
    """
    if os.environ.get("ANTHROPIC_API_KEY"):
        logger.info("synapse.auto_llm: using Anthropic Haiku 4.5")
        return from_anthropic(model="claude-haiku-4-5-20251001")
    if os.environ.get("OPENAI_API_KEY"):
        logger.info("synapse.auto_llm: using OpenAI gpt-4o-mini")
        return from_openai(model="gpt-4o-mini")
    if os.environ.get("GEMINI_API_KEY") or os.environ.get("GOOGLE_API_KEY"):
        from synapse.adapters.hosted.gemini_adapter import GeminiAdapter
        logger.info("synapse.auto_llm: using Gemini 2.5 Flash")
        return GeminiAdapter(model="gemini-2.5-flash")
    if os.environ.get("OLLAMA_HOST") or _ollama_local_running():
        from synapse.adapters.local.ollama_adapter import OllamaAdapter
        logger.info("synapse.auto_llm: using local Ollama (llama3.1:8b)")
        return OllamaAdapter(model="llama3.1:8b")
    raise RuntimeError(
        "synapse.auto_llm: no LLM provider keys found in environment. "
        "Set ANTHROPIC_API_KEY / OPENAI_API_KEY / GEMINI_API_KEY, run a "
        "local Ollama, or call synapse.set_llm(...) explicitly."
    )


def _ollama_local_running() -> bool:
    try:
        import httpx  # type: ignore
        r = httpx.get("http://localhost:11434/api/tags", timeout=0.5)
        return r.status_code == 200
    except Exception:
        return False


# ---------------------------------------------------------------------------
# Internal: minimal InferenceAdapter shim for LangChain-wrapped LLMs.
# Synapse's L3-router and audit-mode scope-inference fallback only need
# a synchronous-style "given prompt, return text" call. We don't reproduce
# the full streaming surface here — just enough for the LLM-mediated
# reasoning paths.
# ---------------------------------------------------------------------------
class _LangChainBridgeAdapter:
    """Best-effort adapter for any LangChain BaseLanguageModel."""

    def __init__(self, llm: Any) -> None:
        from synapse.messages import BackendCapabilities

        self._llm = llm
        self.capabilities = BackendCapabilities(
            backend_id=f"langchain:{type(llm).__name__}",
            tier="hosted",
            supports_midstream_inject=False,
            supports_partial_preservation=False,
            prompt_cache_available=False,
            supports_thinking=False,
            multi_tenant_isolation="none",
        )

    async def generate(
        self,
        messages: list[dict[str, Any]],
        *,
        max_tokens: int = 1024,
        temperature: float = 0.0,
        **_: Any,
    ) -> str:
        """Synapse-internal shorthand: prompt → text."""
        from langchain_core.messages import HumanMessage, SystemMessage, AIMessage

        lc_messages = []
        for m in messages:
            role = m.get("role")
            content = m.get("content", "")
            if role == "system":
                lc_messages.append(SystemMessage(content=content))
            elif role == "assistant":
                lc_messages.append(AIMessage(content=content))
            else:
                lc_messages.append(HumanMessage(content=content))

        result = await self._llm.ainvoke(lc_messages)
        return getattr(result, "content", str(result))

    # Placeholder streaming methods — Synapse's L3 + audit paths only call
    # .generate(). Streaming raises UnsupportedCapability with a clear message.
    async def start_stream(self, *args, **kwargs):  # pragma: no cover
        from synapse.adapters.base import UnsupportedCapability
        raise UnsupportedCapability(
            "LangChain bridge supports .generate() only. Use a native adapter "
            "(AnthropicAdapter / OpenAIAdapter) for streaming."
        )

    def read_tokens(self, handle):  # pragma: no cover
        async def _empty():
            if False:
                yield None
        return _empty()

    async def inject_and_continue(self, handle, injection, instruction=""):  # pragma: no cover
        from synapse.adapters.base import UnsupportedCapability
        raise UnsupportedCapability("LangChain bridge does not support inject_and_continue.")

    async def cancel(self, *args, **kwargs):  # pragma: no cover
        return None


class _LiteLLMBridgeAdapter:
    def __init__(self, model: str, api_key: Optional[str], api_base: Optional[str], extra: dict) -> None:
        from synapse.messages import BackendCapabilities
        try:
            import litellm  # type: ignore[import-not-found]
        except ImportError as e:
            raise RuntimeError(
                "litellm not installed. `pip install litellm`."
            ) from e
        self._litellm = litellm
        self._model = model
        self._api_key = api_key
        self._api_base = api_base
        self._extra = extra
        self.capabilities = BackendCapabilities(
            backend_id=f"litellm:{model}",
            tier="hosted",
            supports_midstream_inject=False,
            supports_partial_preservation=False,
            prompt_cache_available=False,
            supports_thinking=False,
            multi_tenant_isolation="none",
        )

    async def generate(
        self,
        messages: list[dict[str, Any]],
        *,
        max_tokens: int = 1024,
        temperature: float = 0.0,
        **_: Any,
    ) -> str:
        kwargs = {
            "model": self._model,
            "messages": messages,
            "max_tokens": max_tokens,
            "temperature": temperature,
            **self._extra,
        }
        if self._api_key:
            kwargs["api_key"] = self._api_key
        if self._api_base:
            kwargs["api_base"] = self._api_base
        resp = await self._litellm.acompletion(**kwargs)
        return resp.choices[0].message.content or ""

    async def start_stream(self, *args, **kwargs):  # pragma: no cover
        from synapse.adapters.base import UnsupportedCapability
        raise UnsupportedCapability("LiteLLM bridge supports .generate() only")

    def read_tokens(self, handle):  # pragma: no cover
        async def _empty():
            if False:
                yield None
        return _empty()

    async def inject_and_continue(self, handle, injection, instruction=""):  # pragma: no cover
        from synapse.adapters.base import UnsupportedCapability
        raise UnsupportedCapability("LiteLLM bridge does not support inject_and_continue.")

    async def cancel(self, *args, **kwargs):  # pragma: no cover
        return None

File: sdk-typescript/src/llm/bridges.ts
/** Bridges from popular vendor clients to Synapse's InferenceAdapter.
 *
 * Each `from*` accepts the vendor's existing client object so the user keeps
 * their model choice, API key, base URL, retry/timeout config, etc., and
 * returns an adapter Synapse can use.
 *
 * These bridges are LIGHTWEIGHT — they expose `.generate()` for Synapse's
 * internal LLM-mediated reasoning paths (scope inference fallback, belief
 * divergence, auto-merge, L3 semantic routing). Streaming methods raise
 * `UnsupportedCapability`; for streaming use a native adapter.
 */
import {
  type InferenceAdapter,
  type StreamHandle,
  type Token,
  UnsupportedCapability,
} from "../adapters/base.js";
import type { BackendCapabilities } from "../types.js";

export interface ChatMessage {
  role: "system" | "user" | "assistant" | string;
  content: string;
}

export interface GenerateOptions {
  messages: ChatMessage[];
  maxTokens?: number;
  temperature?: number;
  [extra: string]: unknown;
}

/** Subset of InferenceAdapter that bridges actually implement.
 *  Exposes `.generate()` in addition to the streaming surface.
 */
export interface BridgeAdapter extends InferenceAdapter {
  generate(opts: GenerateOptions): Promise<string>;
}

// ---------------------------------------------------------------------------
// Shared no-op streaming surface for bridge adapters
// ---------------------------------------------------------------------------
abstract class BaseBridgeAdapter implements BridgeAdapter {
  abstract capabilities: BackendCapabilities;
  abstract generate(opts: GenerateOptions): Promise<string>;

  async startStream(
    _messages: Array<Record<string, unknown>>,
    _params?: Record<string, unknown>,
  ): Promise<StreamHandle> {
    throw new UnsupportedCapability(
      `${this.constructor.name} supports .generate() only. ` +
        "Use a native adapter (e.g. AnthropicAdapter / OpenAIAdapter) for streaming.",
    );
  }

  readTokens(_handle: StreamHandle): AsyncIterable<Token> {
    return (async function* () {
      // empty
    })();
  }

  async injectAndContinue(
    _handle: StreamHandle,
    _injection: string,
    _instruction?: string,
  ): Promise<StreamHandle> {
    throw new UnsupportedCapability(
      `${this.constructor.name} does not support inject_and_continue.`,
    );
  }

  async cancel(_handle: StreamHandle): Promise<string> {
    return "";
  }
}

/** Dynamic-import indirection — hides the specifier from tsc's static analysis
 *  so optional peer deps (`@anthropic-ai/sdk`, `openai`, `@langchain/core`)
 *  don't need to be installed at typecheck time. Resolves at runtime.
 */
const dynamicImport: (specifier: string) => Promise<any> = new Function(
  "s",
  "return import(s)",
) as (specifier: string) => Promise<any>;

function bridgeCapabilities(backendId: string): BackendCapabilities {
  return {
    backend_id: backendId,
    tier: "hosted",
    supports_midstream_inject: false,
    supports_partial_preservation: false,
    is_reasoning_model: false,
    prompt_cache_available: false,
    avg_overhead_per_signal: 0,
    multi_tenant_isolation: "none",
  };
}

// ---------------------------------------------------------------------------
// Anthropic
// ---------------------------------------------------------------------------
export interface FromAnthropicOptions {
  /** A pre-configured `@anthropic-ai/sdk` client (sync or async). Optional. */
  client?: unknown;
  model?: string;
  apiKey?: string;
}

class AnthropicBridgeAdapter extends BaseBridgeAdapter {
  override capabilities: BackendCapabilities;
  private readonly model: string;
  private readonly apiKey: string | undefined;
  private _client: unknown;

  constructor(opts: FromAnthropicOptions) {
    super();
    this.model = opts.model ?? "claude-haiku-4-5-20251001";
    this.apiKey = opts.apiKey;
    this._client = opts.client ?? null;
    this.capabilities = bridgeCapabilities(`anthropic:${this.model}`);
  }

  private async ensureClient(): Promise<{ messages: { create: Function } }> {
    if (this._client !== null && this._client !== undefined) {
      return this._client as { messages: { create: Function } };
    }
    const mod: any = await dynamicImport("@anthropic-ai/sdk").catch((e) => {
      throw new Error(
        "@anthropic-ai/sdk not installed. `npm install @anthropic-ai/sdk`. " +
          `Original error: ${(e as Error).message}`,
      );
    });
    const Anthropic = mod.default ?? mod.Anthropic ?? mod;
    this._client = new Anthropic(
      this.apiKey !== undefined ? { apiKey: this.apiKey } : {},
    );
    return this._client as { messages: { create: Function } };
  }

  override async generate(opts: GenerateOptions): Promise<string> {
    const client = await this.ensureClient();
    const sys = opts.messages
      .filter((m) => m.role === "system")
      .map((m) => m.content)
      .join("\n");
    const nonSystem = opts.messages
      .filter((m) => m.role !== "system")
      .map((m) => ({
        role: m.role === "assistant" ? "assistant" : "user",
        content: m.content,
      }));
    const resp: any = await client.messages.create({
      model: this.model,
      max_tokens: opts.maxTokens ?? 1024,
      temperature: opts.temperature ?? 0,
      ...(sys ? { system: sys } : {}),
      messages: nonSystem,
    });
    const block = resp?.content?.[0];
    if (block && typeof block.text === "string") return block.text;
    return "";
  }
}

export function fromAnthropic(opts: FromAnthropicOptions = {}): BridgeAdapter {
  return new AnthropicBridgeAdapter(opts);
}

// ---------------------------------------------------------------------------
// OpenAI
// ---------------------------------------------------------------------------
export interface FromOpenAIOptions {
  client?: unknown;
  model?: string;
  apiKey?: string;
  baseURL?: string;
}

class OpenAIBridgeAdapter extends BaseBridgeAdapter {
  override capabilities: BackendCapabilities;
  private readonly model: string;
  private readonly apiKey: string | undefined;
  private readonly baseURL: string | undefined;
  private _client: unknown;

  constructor(opts: FromOpenAIOptions) {
    super();
    this.model = opts.model ?? "gpt-4o-mini";
    this.apiKey = opts.apiKey;
    this.baseURL = opts.baseURL;
    this._client = opts.client ?? null;
    this.capabilities = bridgeCapabilities(`openai:${this.model}`);
  }

  private async ensureClient(): Promise<{ chat: { completions: { create: Function } } }> {
    if (this._client !== null && this._client !== undefined) {
      return this._client as { chat: { completions: { create: Function } } };
    }
    const mod: any = await dynamicImport("openai").catch((e) => {
      throw new Error(
        "openai not installed. `npm install openai`. " +
          `Original error: ${(e as Error).message}`,
      );
    });
    const OpenAI = mod.default ?? mod.OpenAI ?? mod;
    const cfg: Record<string, unknown> = {};
    if (this.apiKey !== undefined) cfg["apiKey"] = this.apiKey;
    if (this.baseURL !== undefined) cfg["baseURL"] = this.baseURL;
    this._client = new OpenAI(cfg);
    return this._client as { chat: { completions: { create: Function } } };
  }

  override async generate(opts: GenerateOptions): Promise<string> {
    const client = await this.ensureClient();
    const resp: any = await client.chat.completions.create({
      model: this.model,
      messages: opts.messages.map((m) => ({ role: m.role, content: m.content })),
      max_tokens: opts.maxTokens ?? 1024,
      temperature: opts.temperature ?? 0,
    });
    return resp?.choices?.[0]?.message?.content ?? "";
  }
}

export function fromOpenAI(opts: FromOpenAIOptions = {}): BridgeAdapter {
  return new OpenAIBridgeAdapter(opts);
}

// ---------------------------------------------------------------------------
// Vercel AI SDK (LanguageModel shape: `doGenerate({ prompt, ... })`)
// ---------------------------------------------------------------------------
export interface VercelLanguageModelLike {
  doGenerate: (args: {
    prompt: Array<{ role: string; content: unknown }>;
    maxTokens?: number;
    temperature?: number;
    [extra: string]: unknown;
  }) => Promise<{ text?: string; content?: unknown; [extra: string]: unknown }>;
  modelId?: string;
  provider?: string;
}

class VercelAIBridgeAdapter extends BaseBridgeAdapter {
  override capabilities: BackendCapabilities;
  private readonly model: VercelLanguageModelLike;

  constructor(model: VercelLanguageModelLike) {
    super();
    this.model = model;
    const id =
      [model.provider, model.modelId].filter(Boolean).join(":") || "vercel-ai";
    this.capabilities = bridgeCapabilities(`vercel-ai:${id}`);
  }

  override async generate(opts: GenerateOptions): Promise<string> {
    // Vercel AI SDK v3+ accepts a `prompt` of typed messages where each message
    // has `role` and `content` (content may be a string or a parts array).
    const prompt = opts.messages.map((m) => ({
      role: m.role,
      content: [{ type: "text", text: m.content }],
    }));
    const result = await this.model.doGenerate({
      prompt,
      ...(opts.maxTokens !== undefined ? { maxTokens: opts.maxTokens } : {}),
      ...(opts.temperature !== undefined ? { temperature: opts.temperature } : {}),
    });
    if (typeof result.text === "string") return result.text;
    // Fallback: some implementations return a `content` array of {type,text} parts.
    if (Array.isArray(result.content)) {
      return result.content
        .map((p: any) => (typeof p?.text === "string" ? p.text : ""))
        .join("");
    }
    return "";
  }
}

/** Wrap a Vercel AI SDK `LanguageModel` (anything exposing `doGenerate`). */
export function fromVercelAI(model: VercelLanguageModelLike): BridgeAdapter {
  if (model === null || typeof model !== "object" || typeof model.doGenerate !== "function") {
    throw new TypeError(
      "fromVercelAI expects a Vercel AI SDK LanguageModel (object with `doGenerate`).",
    );
  }
  return new VercelAIBridgeAdapter(model);
}

// ---------------------------------------------------------------------------
// LangChain.js
// ---------------------------------------------------------------------------
export interface LangChainJSChatModelLike {
  invoke: (
    messages: Array<unknown>,
    options?: Record<string, unknown>,
  ) => Promise<{ content?: unknown; text?: unknown; [extra: string]: unknown }>;
  _llmType?: () => string;
  _modelType?: () => string;
}

class LangChainJSBridgeAdapter extends BaseBridgeAdapter {
  override capabilities: BackendCapabilities;
  private readonly llm: LangChainJSChatModelLike;

  constructor(llm: LangChainJSChatModelLike) {
    super();
    this.llm = llm;
    const id =
      (typeof llm._llmType === "function" && llm._llmType()) ||
      (typeof llm._modelType === "function" && llm._modelType()) ||
      llm.constructor?.name ||
      "model";
    this.capabilities = bridgeCapabilities(`langchain-js:${id}`);
  }

  override async generate(opts: GenerateOptions): Promise<string> {
    // Try to use proper LangChain message classes if available; otherwise
    // fall back to the plain {role, content} shape that most LangChain.js
    // chat models also accept.
    let lcMessages: unknown[];
    try {
      const mod: any = await dynamicImport("@langchain/core/messages");
      lcMessages = opts.messages.map((m) => {
        if (m.role === "system") return new mod.SystemMessage(m.content);
        if (m.role === "assistant") return new mod.AIMessage(m.content);
        return new mod.HumanMessage(m.content);
      });
    } catch {
      lcMessages = opts.messages.map((m) => ({ role: m.role, content: m.content }));
    }

    const result: any = await this.llm.invoke(lcMessages);
    if (typeof result?.content === "string") return result.content;
    if (Array.isArray(result?.content)) {
      return result.content
        .map((p: any) => (typeof p?.text === "string" ? p.text : ""))
        .join("");
    }
    if (typeof result?.text === "string") return result.text;
    return String(result ?? "");
  }
}

export function fromLangChainJS(llm: LangChainJSChatModelLike): BridgeAdapter {
  if (llm === null || typeof llm !== "object" || typeof llm.invoke !== "function") {
    throw new TypeError(
      "fromLangChainJS expects a LangChain.js BaseChatModel (object with `invoke`).",
    );
  }
  return new LangChainJSBridgeAdapter(llm);
}

// ---------------------------------------------------------------------------
// Auto-detect (best-effort) — picks the first provider it finds keys for.
// Order: Anthropic → OpenAI → Vercel AI (any AI_* key) → Ollama (local).
// ---------------------------------------------------------------------------
export interface AutoLlmOptions {
  /** Override `process.env` (mostly for tests). */
  env?: Record<string, string | undefined>;
}

export function autoLlm(opts: AutoLlmOptions = {}): BridgeAdapter {
  const env = opts.env ?? (typeof process !== "undefined" ? process.env : {});
  if (env["ANTHROPIC_API_KEY"]) {
    return fromAnthropic({ model: "claude-haiku-4-5-20251001" });
  }
  if (env["OPENAI_API_KEY"]) {
    return fromOpenAI({ model: "gpt-4o-mini" });
  }
  if (env["AI_GATEWAY_API_KEY"] || env["VERCEL_AI_API_KEY"]) {
    // Vercel AI SDK requires a user-provided LanguageModel — we can't fabricate
    // one without the SDK's provider modules. Emit a clear error pointing the
    // user at fromVercelAI(model).
    throw new Error(
      "synapse.autoLlm: detected Vercel AI env keys but Vercel AI requires an " +
        "explicit LanguageModel. Call synapse.fromVercelAI(model) yourself.",
    );
  }
  if (env["OLLAMA_HOST"]) {
    return fromOpenAI({
      model: "llama3.1:8b",
      baseURL: `${env["OLLAMA_HOST"]}/v1`,
      apiKey: "ollama",
    });
  }
  throw new Error(
    "synapse.autoLlm: no LLM provider keys found in environment. " +
      "Set ANTHROPIC_API_KEY / OPENAI_API_KEY / OLLAMA_HOST, or call " +
      "synapse.setLlm(...) explicitly.",
  );
}
